pkg/iam_roles/iam_role.go

```diff
@@ -0,0 +1,134 @@
+package iam_role
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"regexp"
+	controllerruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// IamRoleConfig : config for AWS IAM role settings for a paasta container
+type IamRoleConfig struct {
+	// +optional
+	IamRoleProvider *string `json:"iam_role_provider,omitempty"`
+	// +optional
+	IamRole *string `json:"iam_role,omitempty"`
+	// PAASTA-16919: remove everything related to fs_group when
+	// https://github.com/aws/amazon-eks-pod-identity-webhook/issues/8 will be
+	// fixed.
+	// +optional
+	FsGroup *int64 `json:"fs_group,omitempty"`
+}
+
+var defaultIamRoleProvider = "kiam"
+var DefaultFsGroup int64 = 65534
+
+var serviceAccountRegex = regexp.MustCompile("[^0-9a-zA-Z]+")
+
+// SetIamRoleConfigDefaults: sets the default values for the AWS IAM role config
+func SetIamRoleConfigDefaults(iamRoleConfig *IamRoleConfig) {
+	if iamRoleConfig.IamRoleProvider == nil {
+		iamRoleConfig.IamRoleProvider = &defaultIamRoleProvider
+	}
+	if iamRoleConfig.FsGroup == nil {
+		iamRoleConfig.FsGroup = &DefaultFsGroup
+	}
+}
+
+// EnsureForIamRole: prepare AWS IAM fole for use
+func EnsureForIamRole(ctx context.Context, client controllerruntimeclient.Client, namespace string, iamRoleConfig *IamRoleConfig) error {
+	// We need to create a service account only for "aws" provider
+	if iamRoleConfig.IamRoleProvider != nil && *iamRoleConfig.IamRoleProvider == "aws" {
+		glog.V(4).Infof("Ensuring service account in %s for iam_role %v exists", namespace, iamRoleConfig.IamRole)
+		if iamRoleConfig.IamRole == nil {
+			return fmt.Errorf("%s/%v: iam_role must be specified when iam_role_provider is set to 'aws'", namespace, iamRoleConfig.IamRole)
+		}
+
+		saName := getServiceAccountNameForIamRole(iamRoleConfig.IamRole)
+
+		// check if service account with this name+namespace already exists
+		glog.V(4).Infof("Looking for service account called %s in namespace %s", saName, namespace)
+		result := &corev1.ServiceAccount{}
+		err := client.Get(ctx,
+			types.NamespacedName{
+				Name:      saName,
+				Namespace: namespace,
+			},
+			result,
+		)
+
+		if err != nil {
+			glog.Errorf("Error while getting service account: %s", err)
+			if errors.IsNotFound(err) {
+				glog.Infof("Service account not found, creating it\n")
```

Comment:
I don't think you need the `\n` on this (https://pkg.go.dev/github.com/golang/glog#Infof)

Reply:
fixed